### PR TITLE
Return true in canCoerce for row types with field name mismatches 

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java [hunk 1]
```diff
@@ -194,14 +194,16 @@ public void testTypeCompatibility()
         assertThat("row(varchar(2))", "row(varchar(5))").hasCommonSuperType("row(varchar(5))").canCoerceFirstToSecondOnly();
 
         assertThat("row(a integer)", "row(a bigint)").hasCommonSuperType("row(a bigint)").canCoerceFirstToSecondOnly();
-        assertThat("row(a integer)", "row(b bigint)").hasCommonSuperType("row(bigint)").cannotCoerceToEachOther();
-        assertThat("row(integer)", "row(b bigint)").hasCommonSuperType("row(bigint)").cannotCoerceToEachOther();
+        assertThat("row(a integer)", "row(b bigint)").hasCommonSuperType("row(bigint)").canCoerceFirstToSecondOnly();
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java]: Why canCoerceFirstToSecondOnly? Isn't the point for these to be coercible to each other?

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 2]
```diff
@@ -357,24 +357,26 @@ private TypeCompatibility typeCompatibilityForRow(RowType firstType, RowType sec
...
-            }
-
-            fields.add(new RowType.Field(name, commonParameterType.get()));
+            // ignore parameter name for firstCanCoerceToSecond
+            firstCanCoerceToSecond &= typeCompatibility.isFirstCanCoerceToSecond();
+            fields.add(new RowType.Field(commonName, commonParameterType));
         }
 
-        return TypeCompatibility.implicit(RowType.from(fields.build()), secondType);
+        return TypeCompatibility.explicit(RowType.from(fields.build()), firstCanCoerceToSecond);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Not sure I understand what "explicit" means in this context.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 3]
```diff
@@ -388,15 +390,17 @@ private TypeCompatibility typeCompatibilityForCovariantParametrizedType(Type fir
             return TypeCompatibility.incompatible();
         }
 
+        boolean firstCanCoerceToSecond = true;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: This seems unrelated to row type. Did you mean to put it in a previous commit?

Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: It is related.

The previous commit is just a refactor. It didn't introduce any semantic change. The assumption that `canCoerce` is implicitly defined by `getCommonSuperTypes` remains.

This commit introduced code that breaks the assumption. And this section of code is to deal with the broken assumption.

I could see the argument for moving this to the previous commit, but I try to keep the previous commit a refactor only.

File: presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java (hunk 1 above)
Comment (haozhun) [reply presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java]: one is integer, the other is bigint.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 2 above)
Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Maybe you can suggest a better name.

Here, `implicit` means `canCoerce` is set implicitly, based on whether `secondType` is equal to the `commonSuperType`. `explicit` means `canCoerce` is set explicitly.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 3 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Ok, maybe change the commit title/description then? It mentions row type only, but this affects parametric types.

File: presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java (hunk 1 above)
Comment (martint) [reply presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java]: Doh! I was so focused on the names that I ignored the field types.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 2 above)
Comment (martint) [reply presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Oh, I see. So the "implicit" method just says leave it up to the TypeCompatibility object to determine, while "explicit" means the caller is making the choice. Yeah, those names are not ideal. Let me think about something better.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 4]
```diff
@@ -615,4 +640,52 @@ public MethodHandle resolveOperator(OperatorType operatorType, List<? extends Ty
...
+        {
+            // Assert that: firstCanCoerceToSecond => commonSuperType.isPresent
+            // The factory API is designed such that this is guaranteed.
+            checkArgument(!firstCanCoerceToSecond || commonSuperType.isPresent());
+
+            this.commonSuperType = commonSuperType;
+            this.firstCanCoerceToSecond = firstCanCoerceToSecond;
+        }
+
+        private static TypeCompatibility explicit(Type commonSuperType, boolean firstCanCoerceToSecond)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: I think part of the problem with the names of these methods is that they are not symmetric. "Incompatible", which directly conveys the notion that types are not compatible with each other, is not the in the same category as "explicit/implicit", which describe the action or mode of constructing the TypeCompatibility object. To make them more coherent, you're need to go with something closer to that concept and name the methods in the vein of "compatibleButNotCoercible", "compatibleAndCoercible", etc. (names provided for illustration purposes, not that I like them)

Comment (haozhun) [reply presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Potentially, I could remove `implicit` all together and have caller always specify to avoid this naming problem.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 5]
```diff
@@ -615,4 +642,47 @@ public MethodHandle resolveOperator(OperatorType operatorType, List<? extends Ty
...
+        {
+            // Assert that: coercible => commonSuperType.isPresent
+            // The factory API is designed such that this is guaranteed.
+            checkArgument(!coercible || commonSuperType.isPresent());
+
+            this.commonSuperType = commonSuperType;
+            this.coercible = coercible;
+        }
+
+        private static TypeCompatibility compatible(Type commonSuperType, boolean firstCanCoerceToSecond)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Rename argument to "coercible" ?
